File: CRISIS/CRISIS/Services/MacroChartInsightEngine.swift
import Foundation

struct MacroChartInsightEngine {
    func makeAnnotation(from series: [MacroSeries]) -> MacroChartAnnotation? {
        let candidates = series.compactMap { annotation(for: $0) }
        guard let top = candidates.sorted(by: { abs($0.delta) > abs($1.delta) }).first else {
            return nil
        }
        return top
    }

    private func annotation(for series: MacroSeries) -> MacroChartAnnotation? {
        guard let latest = series.points.last,
              let previous = series.points.dropLast().last else {
            return nil
        }
        let delta = latest.value - previous.value
        let direction = delta >= 0 ? "gestiegen" : "gefallen"
        let descriptor = "\(series.kind.title) ist \(direction) um \(delta.formatted(.number.precision(.fractionLength(1))))\(series.kind.unit) seit \(previous.year)."
        return MacroChartAnnotation(
            indicator: series.kind,
            focusYear: latest.year,
            delta: delta,
            message: descriptor
        )
    }
}

struct MacroChartAnnotation: Equatable {
    let indicator: MacroIndicatorKind
    let focusYear: Int
    let delta: Double
    let message: String

    var symbol: String {
        delta >= 0 ? "arrow.up.right.circle.fill" : "arrow.down.right.circle.fill"
    }
}

File: android/core-domain/src/main/java/de/vibecode/crisis/core/domain/MacroChartInsightEngine.kt
package de.vibecode.crisis.core.domain

import de.vibecode.crisis.core.model.MacroSeries

class MacroChartInsightEngine {
    fun makeAnnotation(series: List<MacroSeries>): MacroChartAnnotation? {
        val candidates = series.mapNotNull { annotationFor(it) }
        return candidates.maxByOrNull { kotlin.math.abs(it.delta) }
    }

    private fun annotationFor(series: MacroSeries): MacroChartAnnotation? {
        val latest = series.points.lastOrNull() ?: return null
        val previous = series.points.dropLast(1).lastOrNull() ?: return null
        val delta = latest.value - previous.value
        val direction = if (delta >= 0) "gestiegen" else "gefallen"
        val message = "${series.kind.title} ist $direction um ${"%.1f".format(delta)}${series.kind.unit} seit ${previous.year}."
        return MacroChartAnnotation(series.kind, latest.year, delta, message)
    }
}

data class MacroChartAnnotation(
    val indicator: de.vibecode.crisis.core.model.MacroIndicatorKind,
    val focusYear: Int,
    val delta: Double,
    val message: String
) {
    val symbol: String
        get() = if (delta >= 0) "arrow.up" else "arrow.down"
}
